audible-dl.1
.\" http://man.openbsd.org/mdoc.7
.Dd July 7, 2022
.Dt AUDIBLE-DL 1
.Os
.Sh NAME
.Nm audible-dl
.Nd A downloader for your Audible audiobook library.
.\"======================================================================
.Sh SYNOPSIS
.Nm audible-dl
.Op Fl h, -help
.Op Fl l, -log
.Op Fl a, -account Ar account
.Op Fl i, -import Ar file.har
.Op Fl s, -single Ar file.aax
.\"======================================================================
.Sh DESCRIPTION
.Pp
.Nm
is a simple command-line utility to create offline archives of your
Audible library as DRM-free .m4b files, wrapping
.Xr ffmpeg 1
in order to crack Audible's DRM.  It supports multiple Audible
accounts and can be used to convert pre-downloaded .aax files.
.\"======================================================================
.Ss Prerequisites
.Pp
In order for
.Nm
to be useful, you need two things.  Firstly, a HAR
.Pq HTTP Archive Format
file containing your Audible authentication cookies which we use to
scrape your account.  You can get this
by logging into
.Lk https://audible.com
in your browser, opening the network tab of the element inspector,
then navigating to
.Lk https://audible.com/library/titles
and right-clicking on the GET request to that page, selecting
.Qq Copy All As HAR .
This can then be pasted into a file.  Secondly, your Audible
activation bytes which are required to crack the DRM on .aax files.
There are multiple ways to get them, the one I had the most luck with
was a dedicated plugin for RainbowCrack; just follow the instructions
on
.Lk https://github.com/inAudible-NG/tables.
.\"======================================================================
.Pp
Technically, only the second step is required if you just want to
convert .aax files to .m4b files, though at that point you might as
well just write a quick
.Xr ffmpeg 1
script instead.
.\"======================================================================
.Pp
Once you have your activation bytes and authentication cookies, add
the former to your config file (see
.Sx audible-dl EXAMPLES )
and import the latter with:
.Bd -literal
    audible-dl -i path/to/cookies.har
.Ed
.\"======================================================================
.Ss Options
.Pp
.Nm
supports a few command-line options:
.Bl -tag -width DS
.It Fl h, -help
Display a quick help message.
.It Fl l, -log
Save a log of the scraper's progress into the file
.Pa .audible-dl-debug.log .
The contents of this file may be useful in bug reports.
.It Fl a, -account Ar account
Some operations like converting a single .aax file or importing
authentication cookies from a .har file require that you specify an
account with which to perform the operation.  The argument should be
the account's
.Em name
field in the config file.  This option may be omitted if you have only
one account set up.
.It Fl i, -import Ar path/to/file.har
Import authentication cookies from a HAR archive into the specified account.
.It Fl s, -single Ar path/to/file.aax
Convert a single .aax file into an .m4b file using the specified account.
.El
.\"======================================================================
.Ss Configuration
.Pp
In order to use
.Nm
a YAML config file must be created.  At the very minimum it must
contain a list named
.Ic accounts
where each entry contains at least a
.Ic name
and a
.Ic bytes
field.  If
.Ev AUDIBLE_DL_ROOT
is unset the file should also contain a
.Ic savedir
field specifying the directory in which to save downloaded
audiobooks.  If that variable is set and
.Ic savedir
specifies a directory, then books are saved into that directory rather
than the one pointed to by the variable.
.Pp
More config options may be added in the future, including naming
rules for downloaded books and the ability to specify things like the
.Ic savedir
on a per-account basis.
.\"======================================================================
.Sh ENVIRONMENT
.Bl -tag -width DS
.It Ev AUDIBLE_DL_ROOT
When set to an existing directory, tell
.Nm
to look for all of its state beneath it.  Downloaded books will be
saved there and temporary and system files will be stored in the
.Pa .audible-dl/
subdirectory.
.It Ev XDG_CONFIG_HOME
By default,
.Nm
looks for its config file, authentication cookies, and list of
downloaded books in the
.Pa audible-dl/
subdirectory.  The appropriate configuration directory is inferred
using Golang's
.Ic os.UserConfigDir()
function which will return something completely different on Mac OS,
Windows, and Plan 9.
.It Ev XDG_CACHE_HOME
By default,
.Nm
stores temporary intermediate files in the
.Pa audible-dl/temp/
directory.  The appropriate cache directory is inferred using Golang's
.Ic os.UserCacheDir()
function and will return different values on Mac OS, Windows, and Plan
9.
.El
.\"======================================================================
.Sh FILES
.Bl -tag -width DS
.It Pa config.yml
The core configuration file.
.It Pa downloaded_books.json
A list of the books that have already been downloaded.  This file
allows the you to organize and rename your audiobooks at your
leisure.
.It Pa [name].cookies.json
Each account's authentication cookies.
.El
.\"======================================================================
.Sh EXAMPLES
.Ss Average use-case
.Pp
Most users will likely want to use
.Nm
to download books in its default run mode with
.Ev AUDIBLE_DL_ROOT
unset.  For someone with a single account, their config file,
.Pa ~/.config/audible-dl/config.yml ,
will look like this:
.Bd -literal
    savedir: "~/Audiobooks/Audible/"
    accounts:
      - name: "Personal"
        bytes: "deadbeef"
.Ed
.Pp
After running
and
.Pa ~/Audiobooks/Audible/
will contain all their books as .m4b files ready to be transferred to
a phone or mp3 player or indexed into an audiobook library browser.
.\"======================================================================
.Ss Keeping all state in a single directory
.Pp
I like to synchronize my media between several machines, so my
preference is to keep everything in a single directory.  In my shell's
rc file I have:
.Bd -literal
    export AUDIBLE_DL_ROOT="$HOME/media/audiobooks/audible"
.Ed
.Pp
In
.Pa ~/media/audiobooks/audible/.audible-dl/config.yml
I have:
.Bd -literal
    accounts:
      - name: "Personal"
        bytes: "deadbeef"
      - name: "Other"
        bytes: "beefdead"
.Ed
.Pp
Note that I have two accounts set up.
.\"======================================================================
.Sh SEE ALSO
.Xr ffmpeg 1
.Xr ffprobe 1
.Rs
.%B tables
.%U https://github.com/inAudible-NG/tables
.Re
.Rs
.%B Cozy
.%U https://cozy.sh
.Re
.Rs
.%B Voice
.%U https://github.com/PaulWoitaschek/Voice
.Re
.Rs
.%B OpenAudible
.%U https://openaudible.org
.Re
.Rs
.%B audiobookshelf
.%U https://www.audiobookshelf.org
.Re
.\"======================================================================
.Sh AUTHORS
.An Thalia Wright Aq Mt ymir@ulthar.xyz
.\"======================================================================
.Sh HOME
.Em https://sr.ht/~thalia/audible-dl
.\"======================================================================
.Sh BUGS
.Pp
As of the writing of this I am not aware of any bugs.  If you find
any, most likely due to a change in Audible's website, please report
them by sending a detailed email to
.Mt ~thalia/audible-dl@lists.sr.ht .
If possible, please attach the
.Pa .audible-dl-debug.html
and
.Pa .audible-dl-debug.log
files as well as your config file, cookie file(s), and downloaded
books file with all personal info censored.
.\"======================================================================
.Sh SECURITY CONSIDERATIONS
.Pp
.Nm
stores your Audible authentication cookies in plain-text json files.
This means that an attacker who gains access to them will be able to
log into your Audible account in the browser.  Ideally, we wouldn't
have to manage sensitive data ourselves and would simply source your
username and password from your system's keychain, but I've found
Audible's login process to be too complex to easily reverse engineer.


audible-dl.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"path/filepath"
)

// If the -l or --log flag is passed, in addition to logging to an
// internal buffer, the scraper will log to the file
// .audible-dl-debug.log.  The scraper runs synchronously so there
// should be no risk of race conditions causing the contents of this
// or the account buffers to be mangled
var logFile *os.File = nil

// Each book is stored in one of these
type Book struct {
	Slug         string   // B002VA9SWS
	Title        string   // "The Hitchhiker's Guide to the Galaxy"
	Series       string   // "The Hitchhiker's Guide to the Galaxy"
	Runtime      string   // "5 hrs and 51 minutes"
	Summary      string   // "Seconds before the Earth is demolished..."
	CoverURL     string   // "https://m.media-amazon.com/..."
	FileName     string   // "TheHitchhikersGuidetotheGalaxy"
	DownloadURL  string   // "https://cds.audible.com/..."
	CompanionURL string   // ""
	Authors      []string // ["Douglas Adams"]
	Narrators    []string // ["Steven Fry"]
	SeriesIndex  int      // 1
}

////////////////////////////////////////////////////////////////////////
//             _                          _       _
//   ___ _ __ | |_ _ __ _   _ _ __   ___ (_)_ __ | |_
//  / _ \ '_ \| __| '__| | | | '_ \ / _ \| | '_ \| __|
// |  __/ | | | |_| |  | |_| | |_) | (_) | | | | | |_
//  \___|_| |_|\__|_|   \__, | .__/ \___/|_|_| |_|\__|
//                      |___/|_|
////////////////////////////////////////////////////////////////////////

func main() {
	account, harpath, aaxpath, savelog := getArgs()
	cfgfile, datadir, tempdir, savedir := getPaths()
	client := MakeClient(cfgfile, tempdir, savedir, datadir)
	client.Validate()

	if savelog {
		var err error
		logFile, err = os.OpenFile(".audible-dl-debug.log",
			os.O_WRONLY|os.O_CREATE, 0644)
		expect(err, "Failed to open log file for writing")
	}

	if harpath != "" {
		client.ImportCookies(account, harpath)
		os.Exit(0)
	}

	if aaxpath != "" {
		m4b := client.ConvertSingleBook(account, aaxpath)
		fmt.Printf("%s: made %s\n", account, filepath.Base(m4b))
		os.Exit(0)
	}

	client.GetCookies()
	client.GetDownloaded()
	client.ScrapeLibrary(account)

	logFile.Close()
}

////////////////////////////////////////////////////////////////////////
//                   _ _ _            _
//   __ _ _   ___  _(_) (_) __ _ _ __(_) ___  ___
//  / _` | | | \ \/ / | | |/ _` | '__| |/ _ \/ __|
// | (_| | |_| |>  <| | | | (_| | |  | |  __/\__ \
//  \__,_|\__,_/_/\_\_|_|_|\__,_|_|  |_|\___||___/
////////////////////////////////////////////////////////////////////////

const helpMessage string = `Usage: audible-dl [-h] [-a ACC] [-i HAR] [-s AAX]

  Scrape your Audible library or convert an AAX file to m4b.
  See audible-dl(1) for more information.

Options:
  -h, --help         Print this message and exit.
  -a, --account NAME Specify an account for the operation.
  -i, --import  HAR  Import login cookies from HAR.
  -s, --single  AAX  Convert the single AAX file specified in AAX.
  -l, --log          Log scraper info to .audible-dl-debug.log
`

const debugScraperMessage string = `I encountered an error while scraping your library.
There are two likely causes of this:

  1. Your authentication cookies for %s have expired.  You can re-import
     them with "audible-dl -i path/to/cookies.har -a %s", see the man
     page for details.
  2. Audible changed the structure of their website.  This is most likely the
     case if the file .audible-dl-debug.html contains a list of books or
     otherwise looks like you were signed in correctly.

If re-importing your cookies doesn't help, please email a bug report to
"~thalia/audible-dl@lists.sr.ht", see the man page for details.
`

// Like Rust's .unwrap() method.
func unwrap(err interface{}) {
	if err != nil {
		log.Fatal(err)
	}
}

// Like Rust's .expect() method.
func expect(err interface{}, why string) {
	if err != nil {
		log.Print(why)
		log.Print(err)
		fmt.Fprintf(os.Stderr, helpMessage)
		os.Exit(1)
	}
}

////////////////////////////////////////////////////////////////////////
//                  _                                      _
//   ___ _ ____   _(_)_ __ ___  _ __  _ __ ___   ___ _ __ | |_
//  / _ \ '_ \ \ / / | '__/ _ \| '_ \| '_ ` _ \ / _ \ '_ \| __|
// |  __/ | | \ V /| | | | (_) | | | | | | | | |  __/ | | | |_
//  \___|_| |_|\_/ |_|_|  \___/|_| |_|_| |_| |_|\___|_| |_|\__|
////////////////////////////////////////////////////////////////////////

// Audible-dl needs to access several locations in the filesystem in
// order to save downloaded books and discover config, auth, and cache
// data.  Typically, it will look for internal files in the
// OS-specific cache and config directories with downloaded books
// stored in a location that the user must specify in the config
// file.  If $AUDIBLE_DL_ROOT points to a directory, all program state
// will live inside it.
func getPaths() (string, string, string, string) {
	var cfgfile, datadir, tempdir, savedir string
	root := os.Getenv("AUDIBLE_DL_ROOT")
	if root != "" {
		cfgfile = root + "/.audible-dl/config.yml"
		tempdir = root + "/.audible-dl/temp/"
		datadir = root + "/.audible-dl/"
		savedir = root + "/"
	} else {
		conf, err := os.UserConfigDir()
		unwrap(err)
		cach, err := os.UserCacheDir()
		unwrap(err)
		cfgfile = conf + "/audible-dl/config.yml"
		tempdir = cach + "/audible-dl/temp/"
		// FIXME: Ideally these would go in XDG_DATA_HOME but
		// golang doesn't have an os.UserDataDir() so we're
		// stricking them in with the config file instead.
		datadir = conf + "/audible-dl/"
		savedir = "" // Read later from config file
	}
	return cfgfile, datadir, tempdir, savedir
}

// Read command-line arguments.
func getArgs() (string, string, string, bool) {
	var a, h, s string
	var l bool
	// FIXME: prevent duplicate flags
	flag.StringVar(&a, "a", "", "")
	flag.StringVar(&h, "i", "", "")
	flag.StringVar(&s, "s", "", "")
	flag.BoolVar(&l, "l", false, "")
	flag.StringVar(&a, "account", "", "")
	flag.StringVar(&h, "import", "", "")
	flag.StringVar(&s, "single", "", "")
	flag.BoolVar(&l, "log", false, "")
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, helpMessage)
	}
	flag.Parse()
	return a, h, s, l
}
